core/trino-main/src/main/java/io/trino/sql/analyzer/AggregationAnalyzer.java

```diff
@@ -175,14 +177,13 @@ private AggregationAnalyzer(
                 .map(expression -> scopeAwareKey(expression, analysis, sourceScope))
                 .collect(toImmutableSet());
 
-        this.columnReferences = analysis.getColumnReferenceFields()
-                .entrySet().stream()
-                .collect(toImmutableMap(Map.Entry::getKey, entry -> entry.getValue().getFieldId()));
+        this.columnReferences = analysis.getColumnReferenceFields();
```

Comment:
this.columnReferences is (indirectly) mutable now

Reply:
BTW looks like this place is the most important change in the commit.
How can we prevent it from being undone by an innocent refactor?

Reply:
> analysis.getColumnReferenceFields()

returns an immutable map

> How can we prevent it from being undone by an innocent refactor?

I don't quite follow why would anybody want to change that. I see no gain in this.

Reply:
> How can we prevent it from being undone by an innocent refactor?

An automated benchmark for planner?

Reply:
That's a good idea long-term. But this exact problem emerges only for quite specific queries. Mine had 80k columns referenced and 556 group by clauses.

Reply:
> > analysis.getColumnReferenceFields()
> 
> returns an immutable map

no. it returns an unmodifiable map, so the caller cannot mutate it, but the map still can be mutated 

> That's a good idea long-term. But this exact problem emerges only for quite specific queries. Mine had 80k columns referenced and 556 group by clauses.

we already have planner benchmarks.
why not add something representing your specific query to the `BenchmarkPlanner`?

Reply:
> no. it returns an unmodifiable map, so the caller cannot mutate it, but the map still can be mutated

You are right. The question is - is this problematic?
- Multithreading is not a problem here.
- The map is strictly incremental, so no "value not found" can be expected as it can only grow during the lifetime of `AggregationAnalyzer`.

> we already have planner benchmarks. why not add something representing your specific query to the BenchmarkPlanner?

I will take a look

Reply:
> > How can we prevent it from being undone by an innocent refactor?
> 
> I don't quite follow why would anybody want to change that. I see no gain in this.

The class misses defensive copy in the constructor so it's natural to add it.

Reply:
I added a comment.

Reply:
What I don't like is that this now relies on `getColumnReferenceFields` returning an immutable map. If the implementation of that method changes, it could indirectly affect this usage without anyone noticing.

Reply:
Maybe we can just make a sanity check, that the map is indeed immutable? If anyone changes the implementation it will fail fast

Reply:
> Maybe we can just make a sanity check, that the map is indeed immutable?

but it's not (it's just an unmodifiable wrapper)

also, you can't verify that it's what it is, since `unmodifiableMap` has no public class you can test for

Reply:
@findepi Unfortunately you are right.
So what are the options?